src/v/coproc/wasm_event_listener.cc

```diff
@@ -0,0 +1,232 @@
+    std::filesystem::path active_path(_active_dir / name.c_str());
+    return ss::file_exists(active_path.string())
+      .then([this,
+             active_path,
+             name = std::move(name),
+             source_code = std::move(source_code)](bool exists) mutable {
+          if (query_action(source_code) == wasm_event_action::remove) {
+              if (!exists) {
+                  vlog(
+                    coproclog.warn,
+                    "Attempt to remove coprocessor {} but it doesn't exist in "
+                    "the active directory {}",
+                    name,
+                    active_path);
+                  return ss::now();
+              }
+              std::filesystem::path mvto(_inactive_dir / name.c_str());
+              vlog(
+                coproclog.info,
+                "Moving coprocessor {} to inactive dir",
+                active_path);
+              return ss::rename_file(active_path.string(), mvto.string());
+          } else {
+              if (exists) {
+                  vlog(
+                    coproclog.warn,
+                    "Attempt to deploy coprocessor with script {} but one "
+                    "already exists: {}",
+                    name,
+                    active_path);
+                  return ss::now();
+              }
+              vlog(
+                coproclog.info,
+                "Writing coprocessor {} to submit dir {}",
+                name,
+                _submit_dir);
+              std::filesystem::path submit_src(_submit_dir / name.c_str());
+              return write_fully(submit_src, std::move(source_code));
+          }
+      });
+}
+
+wasm_event_listener::wasm_event_listener(std::filesystem::path data_dir)
+  : _client(
+    std::vector<unresolved_address>{unresolved_address("127.0.0.1", 9092)})
+  , _wasm_root(data_dir.parent_path() / "coprocessor")
+  , _active_dir(_wasm_root / "active")
+  , _submit_dir(_wasm_root / "submit")
+  , _inactive_dir(_wasm_root / "inactive") {}
+
+ss::future<> wasm_event_listener::start() {
+    if (_gate.is_closed() || _gate.get_count() != 0) {
+        throw std::logic_error(
+          "Attempted to start() the wasm_event_notifier run "
+          "loop, after it has already been started or closed");
+    }
+    return storage::directories::initialize(_active_dir.string())
+      .then([this] {
+          return storage::directories::initialize(_inactive_dir.string());
+      })
+      .then([this] {
+          return storage::directories::initialize(_submit_dir.string());
+      })
+      .then([this] {
+          try {
+              (void)ss::with_gate(_gate, [this] {
+                  return _client.connect().then([this] {
+                      return ss::do_until(
+                        [this] { return _abort_source.abort_requested(); },
+                        [this] { return do_start(); });
+                  });
+              });
+          } catch (const ss::gate_closed_exception& gce) {
```

Comment:
1. for the future, you can use `if (_gate.is_closed())` instead of catching the exception here to avoid more levels of indentation.

2. to me catching this here seems like it might be masking a potential bug race condition. if the only place that the gate is closed is in `::stop` then it looks like this case would only occur when `::start` and `::stop` are racing with each other. is that the correct scenario? can you write a comment about the race in the code because it seems like that would be uncommon.

Reply:
> this pr looks great! nice work. i think its ready to go pending question about double closing an open file, and the race when catching the closed gate.
> 
> also: can you make a ticket if it doesn't already exist to add configuration changes into the kafka client so we don't need to set the retries the way we did here globally?

done! https://github.com/vectorizedio/redpanda/issues/536

Reply:
what's the resolution on 2?

Reply:
I left a comment on line 119

Reply:
if stop is called after start, how would that trigger this condition? wouldn't it have to be called _concurrently_, and if that is the case, we need to say what that race is--because it is either an important and uncommon race to be aware of or the race is a bug.

Reply:
Because in start() does not return the future that with_gate() returns, so yes entering this with_gate and calling stop() (which calls gate.close()) can happen concurrently. I can re-word the comment